File: python/164_最大间距_hard.py
class Solution:
    def maximumGap(self, nums: List[int]) -> int:
        N = len(nums)
        if N<2:
            return 0
        buff = [0]*N
        exp = 1
        max_num = max(nums)
        while(max_num>=exp):
            cnt = [0]*10
            for num in nums:
                digit = (num//exp) % 10
                cnt[digit] += 1
            for i in range(0, 9):
                cnt[i+1] += cnt[i]
            for i in range(N-1, -1, -1):
                digit = (nums[i]//exp) % 10
                buff[cnt[digit]-1] = nums[i]
                cnt[digit] -= 1
            # 注意python 列表的复制要用切片，否则会指向同一个内存空间
            nums = buff[:]
            exp *= 10
        ans = -1
        for i in range(0, N-1):
            ans = max(ans, nums[i+1]-nums[i])
        return ans

File: cpp/164_最大间距_hard.cpp
//要求线性时间复杂度，因此采用基数排序

class Solution {
public:
    int maximumGap(vector<int>& nums) {
        if (nums.size()<2)
            return 0;
        int N = nums.size();
        int max_num = *max_element(nums.begin(), nums.end());
        vector<int> buff(N);
        int exp = 1;
        while(max_num>=exp){
            vector<int> cnt(10, 0);
            for(int i=0; i<N; i++){
                int digit = (nums[i]/exp) % 10;
                cnt[digit]++;
            }
            for(int i=0; i<10-1; i++)
                cnt[i+1] += cnt[i];
            for(int i=N-1; i>-1; i--){
                int digit = (nums[i]/exp) % 10;
                buff[cnt[digit]-1] = nums[i];
                cnt[digit]--;
            }
            copy(buff.begin(), buff.end(), nums.begin());
            exp *= 10;
        }
        int res = INT_MIN;
        for(int i=0; i<N-1; i++)
            res = max(nums[i+1]-nums[i], res);
        return res;
    }
};